File: web/src/lib/remote/zeigerform.ts
/**
 * Fernsteuerung — **die Form des Host-Zeigers**, vom Host zum Steuernden.
 *
 * Das Cursor-Echo nimmt den Zeiger des Hosts aus dem Bild, damit der Steuernde
 * nur seinen eigenen, verzögerungsfreien sieht
 * (`streaming/win-hq-sidecar/src/capture/cursorsteuerung.rs`). Mit ihm
 * verschwindet aber alles, was ein Zeiger sonst noch erzählt: I-Balken über
 * Text, Doppelpfeil an Kanten, Hand über Verweisen, Wartekringel. Dieses Modul
 * holt genau das zurück — als **Namen**, nicht als Bild:
 *
 * * **Host:** hört die `remote_pointer`-Meldungen seiner Sidecars mit (dieselbe
 *   Brücke wie alle Sidecar-Ereignisse) und schickt jeden Wechsel als
 *   `remote_signal` der Art 'zeiger' an den Steuernden.
 * * **Steuernder:** reicht ihn ins Player-Fenster, wo winit die Form auf den
 *   lokalen Zeiger setzt (`streaming/pulse-player/src/app/eingabe.rs`).
 *
 * ## Warum bevorzugt Namen und nicht Pixel
 *
 * Ein Name kostet ein paar Byte je Wechsel, und gezeichnet wird weiter der
 * lokale Zeiger — also ohne Verzögerung, in der Zeigergröße und dem Thema des
 * Steuernden, und **plattformübergreifend**: winit übersetzt dieselbe
 * CSS-Namensliste unter Windows in `IDC_*`, unter macOS in `NSCursor`, unter
 * Linux in die Namen des installierten Zeiger-Themas. Ein Linux-Rechner, der
 * einen Windows-Rechner steuert, bekommt so seinen eigenen I-Balken.
 *
 * Nur trägt der Name allein die dreizehn Formen, die Windows selbst mitbringt.
 * Die Rasierklinge einer Schnittanwendung, der Werkzeugzeiger einer
 * Bildbearbeitung, der Achsenzeiger eines 3D-Programms fielen früher alle auf
 * `default`. Für die schickt der Host deshalb zusätzlich das **Bild**
 * (`streaming/win-hq-sidecar/src/remote_input/zeigerpixel.rs`), und dieses Modul
 * reicht es mit durch. Der Name bleibt trotzdem immer dabei — er ist der
 * Rückfall, wenn das Bild fehlt oder sich drüben nicht bauen lässt.
 *
 * ## Warum überhaupt ein Filter hier
 *
 * Die Form kommt über den Gateway vom Gegenüber und ist damit Fremdeingabe wie
 * jede andere. Der Player deutet sie zwar selbst und kennt Unbekanntes nicht —
 * aber was nicht auf der Liste steht, hat auch nichts im IPC zum Hauptprozess
 * verloren. **Die Liste ist an drei Stellen dieselbe** und muss synchron
 * bleiben: hier, im Sidecar (`remote_input/zeigerform.rs::abbildung`) und im
 * Player (`app/zeigerform.rs`). Die beiden Rust-Enden hält je ein Test fest;
 * hier trägt sie der Typ [`Zeigerform`] — ein hier erfundener Name fiele erst
 * beim Player auf, und zwar als wortloser Standardpfeil.
 *
 * ## Der Rückfall, wenn der Host die Form gar nicht mehr kennt
 *
 * macOS liest die Zeigerform über eine von Apple **abgekündigte** Abfrage. Fällt
 * sie eines Tages aus, legt der Mac seinen Zeiger zurück ins Videobild und
 * meldet das als `remote_signal` der Art 'zeiger_im_bild' ([`_signalImBild`]);
 * der Player blendet dann seinen lokalen Zeiger aus, damit nicht zwei zu sehen
 * sind. Der Host-Zeiger ist danach formrichtig, aber der Hand um die
 * Übertragungszeit hinterher — schlechter, nicht kaputt. Die Entscheidung dazu
 * steht in [`./zeigerImBild`].
 *
 * **Die Prüfung des BILDES steht nebenan** ([`./zeigerbildPruefung`]), und
 * zwar, damit sie ausführbar ist: dieses Modul importiert `./sidecarInput` zur
 * Laufzeit, und ein solcher Import macht eine Datei für Nodes Testläufer
 * unerreichbar (er löst erweiterungslose Pfade nicht auf). Der Test dort prüft
 * gegen den Prüfstein `streaming/zeigerbild-formen.json`, also gegen die
 * Formen, die der SENDER erzeugt — nicht gegen ausgedachte.
 */

import type { RemoteSignalKind } from '$lib/ws/handlers/types';
import { aufSidecarEreignisse } from './sidecarInput';
import { pruefeBild, type Zeigerbild } from './zeigerbildPruefung';
import { ZeigerImBild, sidecarMeldungImBild } from './zeigerImBild';

export type { Zeigerbild };

type SignalSender = (kind: RemoteSignalKind, data: unknown) => boolean;
/**
 * Wohin Form und Bild beim Steuernden fließen. `imBild` ist der Rückfall: gilt
 * er, blendet der Player seinen lokalen Zeiger ganz aus und der Host-Zeiger
 * reitet im Videobild mit (s. [`./zeigerImBild`]). Die Form geht trotzdem mit —
 * sie gilt wieder, sobald der Rückfall endet.
 */
type Senke = (form: Zeigerform, bild?: Zeigerbild, imBild?: boolean) => void;

/**
 * Die Formen, die über die Leitung dürfen — Namen aus der CSS-Zeigerliste, die
 * winit auf allen drei Plattformen kennt.
 */
const FORMEN = [
  'default',
  'text',
  'pointer',
  'wait',
  'progress',
  'crosshair',
  'help',
  'not-allowed',
  'ew-resize',
  'ns-resize',
  'nwse-resize',
  'nesw-resize',
  'move',
] as const;

export type Zeigerform = (typeof FORMEN)[number];

/** Was gilt, solange nichts Gültiges gemeldet wurde. */
const VORGABE: Zeigerform = 'default';

function istForm(wert: unknown): wert is Zeigerform {
  return typeof wert === 'string' && (FORMEN as readonly string[]).includes(wert);
}

/**
 * Nach wie vielen Millisekunden ohne Wechsel dieselbe Form erneut hinausgeht.
 *
 * Der Sidecar wiederholt sie je Sekunde; ohne diese Auffrischung hier wäre das
 * wirkungslos, denn der Wechselfilter unten verschluckte die Wiederholung. Und
 * ohne Wiederholung bliebe ein Wechsel, den der Sekundendeckel des Gateways
 * still verworfen hat, für den Rest der Sitzung verloren — der Steuernde
 * behielte den I-Balken, während der Host längst wieder auf dem Desktop steht.
 * Etwas unter einer Sekunde, damit die Auffrischung des Sidecars nicht
 * regelmäßig knapp danebenfällt.
 */
const AUFFRISCH_MS = 900;

class RemoteZeigerform {
  #rolle: 'controller' | 'host' | null = null;
  #sendSignal: SignalSender | null = null;
  /** Abmelder der Sidecar-Ereignisse (nur beim Host gesetzt). */
  #abmelden: (() => void) | null = null;
  /** Wohin die Form beim Steuernden geht (Player-Fenster) — best-effort. */
  #senke: Senke | null = null;
  /** Zuletzt gemeldete bzw. gesetzte Form. */
  #form: Zeigerform = VORGABE;
  /** Kennung des zuletzt gemeldeten bzw. gesetzten Bildes, '' = keines. */
  #bildId = '';
  /**
   * Das zuletzt gesetzte Bild — nur beim Steuernden, nur zum Nachliefern an ein
   * Fenster, das sich später anhängt (s. [`setSenke`]). Gehalten wird die
   * **Vollform**, sonst könnte das neue Fenster nichts damit anfangen: sein
   * Player hat noch keinen Vorrat, in den eine blosse Kennung greifen könnte.
   */
  #bild: Zeigerbild | undefined;
  /** Wann der Host zuletzt gesendet hat (`Date.now()`), 0 = noch nie. */
  #gesendetMs = 0;
  /**
   * Wann zuletzt ein **vollständiges** Bild hinausging (`Date.now()`).
   *
   * **Getrennt von [`#gesendetMs`], und aus demselben Grund wie `bild_takte`
   * im Sidecar:** ein gemeinsamer Zähler fällt bei jeder Meldung, und dann
   * frischt bei einem Zeiger, der öfter als der Auffrischtakt wechselt,
   * überhaupt nichts mehr auf — also gerade dort nicht, wo die Heilung
   * gebraucht wird (beim Fahren über eine Timeline wechselt der Zeiger
   * mehrmals je Sekunde).
   */
  #vollstaendigMs = 0;
  /**
   * Steht der Host-Zeiger gerade im Videobild statt in einer Formmeldung?
   * Eigener Baustein, weil dort auch das Zurücksetzen beim Sitzungsende sitzt
   * und geprüft wird (s. [`./zeigerImBild`]).
   */
  #imBild = new ZeigerImBild();

  /** Mit dem Übergang der Sitzung nach 'active' rufen — wie `remoteP2P.start`. */
  start(rolle: 'controller' | 'host', sendSignal: SignalSender): void {
    this.stop();
    this.#rolle = rolle;
    this.#sendSignal = sendSignal;
    if (rolle !== 'host') return;
    this.#abmelden = aufSidecarEreignisse((ev) => this.#vomSidecar(ev));
  }

  /** Sitzungsende — der eine Ausgang, wie bei `remoteP2P.stop`. */
  stop(): void {
    // Den Zeiger des Steuernden zurückgeben, bevor die Senke fällt: sonst
    // behielte sein Fenster die letzte Form der Sitzung. Der Player setzt beim
    // Ende der Erfassung ebenfalls zurück — doppelt, weil die beiden Wege
    // (Sitzungsende, Fenster zu) nicht immer in derselben Reihenfolge laufen.
    //
    // **Der Rückfall gehört unbedingt dazu.** Lief er, ist der lokale Zeiger
    // des Steuernden gerade AUSGEBLENDET — bliebe er das, säße der Nutzer nach
    // dem Ende der Fernsteuerung ohne Zeiger vor seinem eigenen Rechner. Das
    // ist der schlimmste denkbare Ausgang dieser Funktion, und deshalb steht
    // die Rückstellung an derselben Stelle wie die der Form, nicht daneben.
    const rueckfallStand = this.#imBild.beenden();
    if (
      this.#rolle === 'controller' &&
      (this.#form !== VORGABE || this.#bildId || rueckfallStand)
    ) {
      this.#senke?.(VORGABE, undefined, false);
    }
    this.#abmelden?.();
    this.#abmelden = null;
    this.#rolle = null;
    this.#sendSignal = null;
    this.#form = VORGABE;
    this.#bildId = '';
    this.#bild = undefined;
    this.#gesendetMs = 0;
    this.#vollstaendigMs = 0;
  }

  /**
   * Wohin die Form beim Steuernden fließt. Wird beim Setzen sofort mit dem
   * aktuellen Stand beliefert — das Player-Fenster hängt sich später an als die
   * Sitzung beginnt, und ohne die Nachlieferung bliebe die erste Form liegen,
   * bis sich zufällig etwas ändert. Gleiche Schiene wie
   * `remoteP2P.setStatusSink`.
   */
  setSenke(senke: Senke | null): void {
    this.#senke = senke;
    if (senke && this.#rolle === 'controller') {
      senke(this.#form, this.#bild, this.#imBild.aktiv);
    }
  }

  /**
   * Ein `remote_signal` der Art 'zeiger' vom Gegenüber (Zuordnung zur Sitzung
   * prüft der Handler).
   *
   * **Nur der Steuernde hört zu.** Der Host ist die Quelle dieser Auskunft; ein
   * 'zeiger' in seine Richtung kann nur aus einem selbstgebauten Client kommen
   * und hat dort nichts zu bestellen.
   */
  _signal(data: unknown): void {
    if (this.#rolle !== 'controller') return;
    if (!data || typeof data !== 'object') return;
    const { form, bild } = data as { form?: unknown; bild?: unknown };
    // Unbekannte Form → Standardpfeil, nicht ignorieren: eine ausgedachte
    // Form soll nicht die letzte gültige stehen lassen. Für das Bild gilt
    // dasselbe: was die Prüfung nicht besteht, gibt es hier nicht.
    const gueltig = istForm(form) ? form : VORGABE;
    const gueltigesBild = pruefeBild(bild);
    const kennung = gueltigesBild?.id ?? '';
    // **Eine Wiederholung MIT Daten geht trotzdem durch.** Sie ist die
    // Auffrischung des Hosts, und sie ist der einzige Weg, auf dem sich ein
    // Bild heilt, das drüben fehlt — etwa weil der Player seinen Vorrat
    // geleert hat, während der Host es weiter für bekannt hält. Ohne diese
    // Ausnahme bliebe der Steuernde bis zum nächsten echten Wechsel beim
    // Standardpfeil. Reine Wiederholungen ohne Daten kosten dagegen nur IPC.
    if (gueltig === this.#form && kennung === this.#bildId && !gueltigesBild?.daten) return;
    this.#form = gueltig;
    this.#bildId = kennung;
    // Nur die Vollform aufheben: ein später angehängtes Fenster kann mit einer
    // blossen Kennung nichts anfangen, sein Player hat noch keinen Vorrat.
    // **Und das Gemerkte muss zur Kennung passen.** Bei der Kurzform `{id}`
    // steht das Bild nicht in der Meldung, sondern im Vorrat des Players —
    // hier bliebe sonst das Bild des VORIGEN Zeigers liegen, während `#bildId`
    // schon auf den neuen zeigt, und `setSenke` reichte einem zweiten
    // Player-Fenster die falsche Form. Passt die Kurzform zum Gemerkten, bleibt
    // es; sonst (und wenn das Bild ganz wegfällt) fällt es weg.
    if (gueltigesBild?.daten) this.#bild = gueltigesBild;
    else if (this.#bild?.id !== kennung) this.#bild = undefined;
    this.#senke?.(gueltig, gueltigesBild, this.#imBild.aktiv);
  }

  /**
   * Ein `remote_signal` der Art 'zeiger_im_bild' vom Gegenüber — der
   * **Rückfall**: der Host kann seine Zeigerform nicht mehr melden und hat
   * seinen Zeiger zurück ins Videobild gelegt.
   *
   * Der Player blendet daraufhin seinen lokalen Zeiger aus; sonst stünden zwei
   * Zeiger im Bild, und der falsche wäre der schnellere. Endet der Rückfall,
   * kommt der lokale zurück und die zuletzt gemeldete Form gilt wieder —
   * deshalb geht sie hier weiter mit.
   *
   * **Nur der Steuernde hört zu**, aus demselben Grund wie bei [`_signal`]: der
   * Host ist die Quelle dieser Auskunft.
   *
   * Die Deutung der Nutzlast steht in [`./zeigerImBild`] und ist bewusst
   * streng — im Zweifel gilt „Zeiger sichtbar".
   */
  _signalImBild(data: unknown): void {
    if (this.#rolle !== 'controller') return;
    // Wiederholungen schluckt der Baustein selbst; nur ein WECHSEL kostet IPC.
    if (!this.#imBild.signal(data)) return;
    this.#senke?.(this.#form, this.#bild, this.#imBild.aktiv);
  }

  // ── Host-Seite ────────────────────────────────────────────────────────────

  #vomSidecar(ev: unknown): void {
    if (this.#rolle !== 'host') return;
    if (!ev || typeof ev !== 'object') return;
    const m = ev as { ev?: unknown; shape?: unknown; bild?: unknown };
    // **Der Rückfall reist über denselben Weg**, aber mit eigener Art. Der
    // Sidecar meldet ihn als `remote_pointer_in_frame`, wenn seine Abfrage
    // nichts hergab und er den Host-Zeiger zurück ins Bild geschaltet hat.
    //
    // **Diese Weiterleitung fehlte bis zum 2026-08-23**, und das war eine
    // Lücke genau zwischen zwei Hälften: der Sidecar meldete, der Player
    // konnte es deuten, der README beschrieb es — nur reichte niemand es
    // weiter, und `#vomSidecar` verwarf die Art wortlos. Der Steuernde hätte
    // seinen Zeiger nie ausgeblendet und immer zwei gesehen. Kein Absturz,
    // keine Meldung: der Rückfall wäre stillschweigend wirkungslos gewesen.
    //
    // Nicht gedeutet wird hier: `aktiv` geht roh hinaus, die Gegenseite
    // entscheidet (`./zeigerImBild`). Ein zweiter Deutungsort wäre ein
    // zweiter Ort, an dem sich die Regel ändern kann.
    const imBild = sidecarMeldungImBild(ev);
    if (imBild !== null) {
      this.#sendSignal?.('zeiger_im_bild', { aktiv: imBild });
      return;
    }
    if (m.ev !== 'remote_pointer') return;
    // Was der eigene Sidecar meldet, ist nicht Fremdeingabe — geprüft wird es
    // trotzdem, damit eine ältere oder neuere Sidecar-Fassung nichts über die
    // Leitung schiebt, das die Gegenseite nicht deuten kann.
    const form = istForm(m.shape) ? m.shape : VORGABE;
    const bild = pruefeBild(m.bild);
    const kennung = bild?.id ?? '';
    const jetzt = Date.now();
    // Der Zeiger ist maschinenweit einer, aber bei mehreren Streams meldet ihn
    // jeder Sidecar-Prozess für sich. Deshalb wird hier zusammengefasst: der
    // Wechsel geht sofort hinaus, die Auffrischung höchstens im Takt von
    // `AUFFRISCH_MS` — sonst ginge sie mit jedem Platz einzeln hinaus. Der
    // Wechsel misst sich an Form UND Bildkennung: zwei Werkzeugzeiger desselben
    // Programms tragen beide den Namen `default`.
    const gleich = form === this.#form && kennung === this.#bildId;
    // Ein vollständiges Bild geht auch dann durch, wenn sich sonst nichts
    // geändert hat: es ist die Auffrischung des Sidecars, und die ist der eine
    // Weg, auf dem ein drüben fehlendes Bild wieder ankommt. Die Sperre gilt
    // dabei getrennt (s. [`#vollstaendigMs`]) — sonst schluckte ein schnell
    // wechselnder Zeiger jede Auffrischung, weil die gemeinsame Sperre nie
    // abläuft.
    const istAuffrischung = Boolean(bild?.daten);
    const sperreMs = istAuffrischung ? this.#vollstaendigMs : this.#gesendetMs;
    if (gleich && jetzt - sperreMs < AUFFRISCH_MS) return;
    this.#form = form;
    this.#bildId = kennung;
    this.#gesendetMs = jetzt;
    if (istAuffrischung) this.#vollstaendigMs = jetzt;
    // Geht die Meldung nicht hinaus (Verbindungs-Blip), wird sie hier NICHT
    // wiederholt: der Sidecar meldet je Sekunde erneut, und die nächste
    // Auffrischung holt es nach. Eine falsche Zeigerform ist zudem der
    // harmloseste Verlust dieser Sitzung — sie kostet Rückmeldung, keine
    // Eingabe.
    this.#sendSignal?.('zeiger', bild ? { form, bild } : { form });
  }
}

export const remoteZeigerform = new RemoteZeigerform();

File: streaming/win-hq-sidecar/src/remote_input/zeigerform.rs
//! Welche **Form** der Zeiger des Hosts gerade hat — die Windows-Hälfte.
//!
//! **Die Buchführung liegt seit dem 2026-08-23 in
//! [`pulse_fernsteuerung::zeigerbuch`]**: was zuletzt gemeldet wurde, was die
//! Gegenseite schon kennt, wann aufgefrischt wird und wie die Nachricht
//! aussieht. Dort steht auch, warum es dieses Merkmal überhaupt gibt (das
//! Cursor-Echo nimmt den Host-Zeiger aus dem Bild und mit ihm alles, was seine
//! Form erzählt) und welche vier Pflichten beim Sidecar bleiben.
//! **Nicht wieder hierher zurückkopieren** — die Datei existiert nur noch
//! einmal.
//!
//! Was hier bleibt, kennt Windows: die Tabelle der Standard-Zeiger
//! ([`abbildung`]), die Abfrage des gerade gezeichneten Zeigers ([`ermitteln`])
//! und die Übersetzung Handle → Name ([`zu_name`]). Dazu die drei Zeilen, die
//! die Kiste an diesen Prozess anschließen: die eine Buchführung samt ihrer
//! Sperre, der Wecker und das Einreihen in den Ereigniskanal.
//!
//! **`CURSOR_SHOWING` bleibt bewusst unausgewertet.** Die Begründung steht
//! einmal, und zwar drüben ([`pulse_fernsteuerung::zeigerbuch`], Abschnitt „Was
//! hier bewusst NICHT entschieden wird") — sie gilt für jeden Sender, nicht nur
//! für diesen. Hier steht nur der Name des Windows-Merkmals, damit die Suche
//! danach hier landet.

use std::sync::Mutex;

use windows::Win32::UI::WindowsAndMessaging::{
    CURSORINFO, GetCursorInfo, HCURSOR, IDC_APPSTARTING, IDC_ARROW, IDC_CROSS, IDC_HAND, IDC_HELP,
    IDC_IBEAM, IDC_NO, IDC_SIZEALL, IDC_SIZENESW, IDC_SIZENS, IDC_SIZENWSE, IDC_SIZEWE, IDC_WAIT,
    LoadCursorW,
};
use windows::core::PCWSTR;

use super::{wache, zeigerpixel};
use pulse_fernsteuerung::zeigerbuch::{Stand, VORGABE, Zeigerbuch};

/// Die eine Buchführung dieses Prozesses.
///
/// Sie liegt hier und nicht in der Kiste: die hält bewusst **keinen** globalen
/// Zustand (s. deren `lib.rs`), damit ihre Tests ohne prozessweite Reihenfolge
/// auskommen. Womit ein Wirt seine eine Buchführung schützt, ist seine Sache.
static BUCH: Mutex<Zeigerbuch> = Mutex::new(Zeigerbuch::LEER);

/// Die Sperre nehmen — auch eine vergiftete, aus demselben Grund wie bei der
/// Sperre der Fernsteuer-Sitzung (`pulse_fernsteuerung::sitzung::Sitzung`):
/// [`zuruecksetzen`] liegt auf jedem Ausstiegsweg der Sitzung und darf an
/// keiner fremden Panik scheitern.
fn sperre() -> std::sync::MutexGuard<'static, Zeigerbuch> {
    BUCH.lock().unwrap_or_else(|e| e.into_inner())
}

/// Die Abbildung Windows-Zeiger → Name. Als Funktion statt als `static`, weil
/// [`PCWSTR`] ein roher Zeiger ist; die Werte selbst sind Konstanten des
/// Betriebssystems, das Aufbauen kostet nichts.
///
/// **Nicht zwischengespeichert:** `LoadCursorW` auf einen Standard-Zeiger ist
/// ein Nachschlagen ohne Ladevorgang, und die Handles wechseln, wenn der Nutzer
/// sein Zeigerschema umstellt (`SetSystemCursor`). Ein gemerktes Handle wäre
/// danach falsch, und die Fernsteuerung meldete für den Rest der Sitzung nur
/// noch [`VORGABE`].
///
/// Ohne Entsprechung bleiben `IDC_UPARROW` und die Personen-/Nadel-Zeiger:
/// dafür gibt es in der CSS-Liste nichts, und etwas Ähnliches zu nehmen wäre
/// geraten. Sie fallen auf [`VORGABE`].
fn abbildung() -> [(PCWSTR, &'static str); 13] {
    [
        (IDC_ARROW, VORGABE),
        (IDC_IBEAM, "text"),
        (IDC_HAND, "pointer"),
        (IDC_WAIT, "wait"),
        (IDC_APPSTARTING, "progress"),
        (IDC_CROSS, "crosshair"),
        (IDC_HELP, "help"),
        (IDC_NO, "not-allowed"),
        (IDC_SIZEWE, "ew-resize"),
        (IDC_SIZENS, "ns-resize"),
        (IDC_SIZENWSE, "nwse-resize"),
        (IDC_SIZENESW, "nesw-resize"),
        (IDC_SIZEALL, "move"),
    ]
}

/// Der gerade gezeichnete System-Zeiger.
fn ermitteln() -> Stand {
    let mut info =
        CURSORINFO { cbSize: std::mem::size_of::<CURSORINFO>() as u32, ..Default::default() };
    if unsafe { GetCursorInfo(&mut info) }.is_err() {
        // Kein Grund für mehr als die Vorgabe: die Abfrage ist Beiwerk, und
        // eine Störung darf weder die Sitzung noch das Protokoll fluten (der
        // Wecker käme 100 ms später mit derselben Zeile wieder) — deshalb wird
        // der Fehler auch nicht ausgegeben.
        return Stand::Name(VORGABE);
    }
    match zu_name(info.hCursor) {
        Some(name) => Stand::Name(name),
        // **Der Zeiger wird bei JEDEM Wecker frisch ausgelesen**, nicht am
        // Handle festgemacht. Windows gibt die Zahl eines freigegebenen Zeigers
        // an den nächsten weiter; wer sie als Ausweis nähme, zeigte irgendwann
        // ein Bild, das zu einem längst verworfenen Zeiger gehört. Das Auslesen
        // ist eine Kopie von wenigen Kilobyte und fällt zehnmal je Sekunde
        // neben der laufenden Bildschirmaufnahme nicht ins Gewicht.
        None => zeigerpixel::bild_holen(info.hCursor).map_or(Stand::Name(VORGABE), Stand::Eigen),
    }
}

/// Ein Zeiger-Handle in einen Namen übersetzen. Getrennt von [`ermitteln`],
/// damit der Vergleich für sich steht — er ist die einzige Stelle, an der
/// dieses Modul etwas behauptet.
///
/// `None` heisst **nicht** „Fehler", sondern „kein Zeiger, den Windows selbst
/// mitbringt" — und damit: hol die Pixel.
fn zu_name(aktuell: HCURSOR) -> Option<&'static str> {
    if aktuell.0.is_null() {
        // Kein Zeiger gesetzt (ausgeblendet). Absichtlich nicht als eigene
        // Form gemeldet — Begründung im Modulkopf. Auch kein Bild: es gibt
        // keines.
        return Some(VORGABE);
    }
    for (kennung, name) in abbildung() {
        if unsafe { LoadCursorW(None, kennung) }.is_ok_and(|h| h == aktuell) {
            return Some(name);
        }
    }
    None
}

/// Der Wecker der Wache (alle 100 ms, aus ihrem eigenen Faden).
///
/// Hier stehen die vier Pflichten beisammen, die
/// [`pulse_fernsteuerung::zeigerbuch`] dem Sender überlässt: der Takt (dieser
/// Wecker), der [`Stand`], die Vorrang-Weiche und das Einreihen außerhalb der
/// eigenen Sperre.
pub(super) fn tick() {
    // Nur während einer Fernsteuerung: der Wecker überlebt das Sitzungsende um
    // bis zu einen Takt, und ohne Steuernden gibt es niemanden, den die Form
    // angeht.
    if !super::fern_aktiv() {
        return;
    }
    // **Bei Vorrang des Hosts die Vorgabe**, nicht die echte Form: der Host
    // führt dann seinen eigenen Zeiger, der wieder im Bild ist
    // (Vorrang-Übergang der Sitzung) — der Steuernde soll nicht mit einem
    // I-Balken dastehen, der zu einer Bewegung gehört, die nicht seine ist. Aus
    // demselben Grund geht dann auch kein Bild hinaus. Und `ermitteln` läuft
    // dabei **gar nicht erst**: es kostet `GetCursorInfo`, bis zu dreizehn
    // `LoadCursorW` und im schlechten Fall das Auslesen der Pixel — während der
    // Host selbst arbeitet.
    let stand = if wache::host_regt_sich() { Stand::Name(VORGABE) } else { ermitteln() };
    // Die Sperre endet mit dieser Anweisung, das Einreihen läuft außerhalb:
    // `emit` reiht zwar nur ein, aber es gibt keinen Grund, einen fremden Kanal
    // unter einer eigenen Sperre anzufassen.
    let nachricht = sperre().nachricht(&stand);
    if let Some(n) = nachricht {
        crate::events::emit(n);
    }
}

/// Sitzungsende: das Buch leeren, damit die nächste Sitzung ihre erste Form
/// wieder in jedem Fall meldet (Begründung bei
/// [`Zeigerbuch::zuruecksetzen`]).
pub(super) fn zuruecksetzen() {
    sperre().zuruecksetzen();
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Ein Nullzeiger ist kein Handle: kein Absturz, sondern die Vorgabe — und
    /// ausdrücklich **kein** Anlass, Pixel zu suchen, denn es gibt keine.
    #[test]
    fn ohne_zeiger_gilt_die_vorgabe() {
        assert_eq!(zu_name(HCURSOR(std::ptr::null_mut())), Some(VORGABE));
    }

    /// Jede Form der Tabelle ist ein Name aus der CSS-Liste, den winit auf der
    /// Gegenseite kennt (`streaming/pulse-player/src/app/zeigerform.rs`). Der Test
    /// hält die beiden Enden zusammen: ein hier erfundener Name käme drüben
    /// wortlos als Standardpfeil an.
    #[test]
    fn alle_formen_sind_bekannte_namen() {
        const BEKANNT: &[&str] = &[
            "default",
            "text",
            "pointer",
            "wait",
            "progress",
            "crosshair",
            "help",
            "not-allowed",
            "ew-resize",
            "ns-resize",
            "nwse-resize",
            "nesw-resize",
            "move",
        ];
        for (_, name) in abbildung() {
            assert!(BEKANNT.contains(&name), "{name} steht nicht auf der Liste des Players");
        }
    }
}
